app/services/advance_order_service.rb

```diff
@@ -0,0 +1,35 @@
+class AdvanceOrderService
+  attr_reader :order
+
+  def initialize(order)
+    @order = order
+  end
+
+  def call(raise_on_error = false)
+    shipping_method_id = order.shipping_method.id if order.shipping_method.present?
+    options = { shipping_method_id: shipping_method_id }
+    raise_on_error ? advance_order!(options) : advance_order(options)
+  end
+
+  private
+
+  def advance_order(options)
+    until order.state == "complete"
```

Comment:
It'd be great to stick to the approach in line 24 for consistency IMO

Reply:
I gave this a try just now, @sauloperez, and confirmed that a completed but later cancelled order would attach an error here (`ActiveRecord::Base#errors`) whereas using `order.completed?` would not. `Spree::Order#completed?` looks at `completed_at`, not the order state.

We can actually remove `advance_order!` in Spree 2.2. (Discussion here: [heere](https://github.com/openfoodfoundation/openfoodnetwork/pull/3560#discussion_r265746751)) Are you okay with waiting until then so we avoid doing a more comprehensive check?